### PR TITLE
adding packed bert squad notebook

### PR DESCRIPTION
**Pull request by arsalanu**
### What does this PR do?

Adds a variant of the notebook question_answering with a step-by-step implementation of packed-BERT for fine-tuning. There are modifications to preprocessing, dataset creation, the model class and postprocessing in comparison to unpacked question-answering, all highlighted in this notebook. I've read through and its ready for review, there could be stuff to add or remove.

## Notes on the notebook
**Question:** We would ideally like to have interleaved training-validation loops, but SQuAD validation accuracy metrics require custom processing for logits with multiple arguments to the `postprocess_qa_predictions `function. IPUTrainingArguments lets you add a `preprocess_logits_for_metrics` function but this only takes `logits `as an argument, with no discernible way to add multiple arguments or even `kwargs` (let me know if this is possible). For this reason I've done the validation process separately (after training) using `.predict()`

I've also added an example showing inference using HF pipelines with a single text input, because of the preset task-specific pipelines I've used an unpacked input with the default 'question answering' pipeline to prove model functionality. We can't yet do packed inference without incorporating preprocessing into the inference pipeline. I will remove this if it's not needed.

### CODE REVIEW

File: notebooks/packedBERT_question_answering.ipynb [hunk 1]
```diff
@@ -0,0 +1,2942 @@
...
+      "Requirement already satisfied: multidict<7.0,>=4.5 in /localdata/arsalanu/popsdk_venvs/poplar_sdk-ubuntu_20_04-3.0.0+1145-1b114aac3a/3.0.0+1145_poptorch/lib/python3.8/site-packages (from aiohttp->datasets->optimum-graphcore->optimum[graphcore]) (6.0.2)\n",
+      "Requirement already satisfied: pytz>=2020.1 in /localdata/arsalanu/popsdk_venvs/poplar_sdk-ubuntu_20_04-3.0.0+1145-1b114aac3a/3.0.0+1145_poptorch/lib/python3.8/site-packages (from pandas->datasets->optimum-graphcore->optimum[graphcore]) (2022.6)\n",
+      "Requirement already satisfied: python-dateutil>=2.8.1 in /localdata/arsalanu/popsdk_venvs/poplar_sdk-ubuntu_20_04-3.0.0+1145-1b114aac3a/3.0.0+1145_poptorch/lib/python3.8/site-packages (from pandas->datasets->optimum-graphcore->optimum[graphcore]) (2.8.2)\n",
+      "Requirement already satisfied: six>=1.5 in /localdata/arsalanu/popsdk_venvs/poplar_sdk-ubuntu_20_04-3.0.0+1145-1b114aac3a/3.0.0+1145_poptorch/lib/python3.8/site-packages (from python-dateutil>=2.8.1->pandas->datasets->optimum-graphcore->optimum[graphcore]) (1.16.0)\n"
+     ]
+    }
+   ],
+   "source": [
+    "! pip install optimum[graphcore]"
```
[the remark this answers is not in the document]
Comment (jimypbr) [reply notebooks/packedBERT_question_answering.ipynb:95]: The correct version of `transformers` will be installed as a dependency of optimum-graphcore so just remove it entirely.

[the remark this answers is not in the document]
Comment (arsalanu) [reply notebooks/packedBERT_question_answering.ipynb:95]: Have to do the git install to use transformers 4.25.1, pip latest is 4.20.1 with optimum-graphcore 0.5.0 - the git install does 0.6.0 and 4.25.1 - changing this in new pr

File: notebooks/packedBERT_question_answering.ipynb [hunk 2]
```diff
@@ -0,0 +1,2942 @@
...
+    "ipu_config_name=\"Graphcore/bert-base-uncased\" # Default Graphcore IPU config initialisation for pre-trained BERT\n",
+    "max_seq_length=384 # The maximum sequence length allowed for sequences in the model.\n",
+    "max_seq_per_pack=3 # The maximum number of sequences that can be packed into a single packed input sequence.\n",
+    "\n",
+    "micro_batch_size_train=1 # Local batch size used during training.\n",
+    "micro_batch_size_eval=8  # Local batch size used during validation.\n",
+    "gradient_accumulation_steps=32 # Gradient accumulation steps for training the model on the IPU.\n",
+    "device_iterations=16 # Device iterations for training the model on the IPU.\n",
+    "replication_factor=1 # Replication of the model across this many sets of IPUs.\n",
+    "pod_type=\"pod4\" # The model is configured to run on a POD with 4 IPUs."
```
[the remark this answers is not in the document]
Comment (arsalanu) [reply notebooks/packedBERT_question_answering.ipynb:319]: modified to use env variables in new pr

File: notebooks/packedBERT_question_answering.ipynb [hunk 3]
```diff
@@ -0,0 +1,2942 @@
...
+    "            ref_inds = []\n",
+    "            for x in strategy:\n",
+    "                ref_ind = np.argwhere(sorted_seqs[0] == x)[-1]\n",
+    "                sorted_seqs[0, ref_ind] = -1\n",
+    "                ref_inds.append(ref_ind)\n",
+    "\n",
+    "            inds = sorted_seqs[1, ref_inds].flatten()\n",
+    "\n",
+    "            # Concatenate the data for each input in the strategy together to create the pack\n",
+    "            input_id_pack       = list(itertools.chain(*[input_ids[x] for x in inds]))\n",
```
[the remark this answers is not in the document]
Comment (arsalanu) [reply notebooks/packedBERT_question_answering.ipynb:1459]: the data that comes into this function is inhomogenous (padding is disabled in the tokenizer intentionally) so is at least directly incompatible with numpy - also the dataset is tokenized into an arrow format, i think you would also need to convert the whole dataset to numpy

to use a numpy here (or torch even) means the tokenizer needs to pad to max length, but then this function would need to unpad the padding and get lengths for each example everytime within the loop - means adding more dynamic indexing for the gain of the flatten+concatenate, felt less intuitive to me when writing but worth testing out - i wont change it for now but i will have a look at making this whole function faster in the future (its also open to multiprocessing)

**Comment by arsalanu** [comment]
Making a new PR with all supported packing tasks in simplified notebooks with accompanying utils/models - closing this PR because its not just the squad task.
